supervisor/supervisor_test.go

```diff
@@ -0,0 +1,211 @@
+// pmm-agent
+// Copyright (C) 2018 Percona LLC
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+package supervisor
+
+import (
+	"context"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/percona/pmm/api/agent"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/percona/pmm-agent/config"
+)
+
+func agentProcessIsExists(t *testing.T, s *Supervisor, agentID uint32) (int, bool) {
+	subAgent, ok := s.agents[agentID]
+	if !ok {
+		t.Errorf("Sub-agent not added to map")
+		return 0, false
+	}
+	pid := subAgent.pid()
+	procExists := processIsExists(pid)
+	return pid, procExists
+}
+
+func processIsExists(pid int) bool {
+	killErr := syscall.Kill(pid, syscall.Signal(0))
+	procExists := killErr == nil
+	return procExists
+}
+
+func checkResponse(t *testing.T, process *agent.SetStateResponse_AgentProcess, disabled bool) {
+	expectedResponse := agent.SetStateResponse_AgentProcess{
+		AgentId:    process.AgentId,
+		ListenPort: process.AgentId + 10000,
+		Disabled:   disabled,
+	}
+	assert.Equal(t, expectedResponse, *process)
+}
+
+func setup() (context.CancelFunc, *Supervisor, []string, []string) {
+	ctx, cancel := context.WithCancel(context.TODO())
+	s := NewSupervisor(ctx, config.Ports{Min: 10001, Max: 20000})
+	arguments := []string{
+		"-web.listen-address=127.0.0.1:{{ .ListenPort }}",
+	}
+	env := []string{
+		`DATA_SOURCE_NAME="pmm:pmm@(127.0.0.1:3306)/pmm-managed-dev"`,
+	}
+	return cancel, s, arguments, env
+}
+
+func TestUpdateStateSimple(t *testing.T) {
+	cancel, s, arguments, env := setup()
+	defer cancel()
+
+	var processes []*agent.SetStateRequest_AgentProcess
+	agentsCount := uint32(5)
+	for i := uint32(1); i <= agentsCount; i++ {
+		processes = append(processes, &agent.SetStateRequest_AgentProcess{
+			AgentId: i,
+			Type:    agent.Type_MYSQLD_EXPORTER,
+			Args:    arguments,
+			Env:     env,
+		})
+	}
+
+	response := s.UpdateState(processes)
+	time.Sleep(1 * time.Second)
```

Comment:
In general, let's try to use smaller sleep values to make tests faster (but also keep them stable).